### PR TITLE
Issue 4159 - Healthcheck code DSBLE0002 not returned on disabled suffix

### LINKED ISSUE #4159: Healthcheck code DSBLE0002 not returned on disabled suffix
Cloned from Pagure issue: https://pagure.io/389-ds-base/issue/51106
- Created at 2020-05-21 22:37:55 by [mreynolds](https://pagure.io/user/mreynolds) (@mreynolds389)
- Assigned to nobody
- Associated bugzillas
  - https://bugzilla.redhat.com/show_bug.cgi?id=1837315

---

Ticket was cloned from Red Hat Bugzilla (product _Red Hat Enterprise Linux 8_): [Bug 1837315](https://bugzilla.redhat.com/show_bug.cgi?id=1837315)

```text
Description of problem:
DSBLE0002 should be returned when suffix is disabled.
But running healthcheck returns only the warning message and it stops checking
other components.

Version-Release number of selected component (if applicable):
389-ds-base-1.4.3.8-2.module+el8.3.0+6591+ebfc9766.x86_64

How reproducible:
Always

Steps to Reproduce:
1. Create a new root suffix:
# ldapmodify -a -p 389 -h localhost -D "cn=Directory manager" -w password
dn: cn="dc=test,dc=com",cn=mapping tree, cn=config
objectclass: top
objectclass: extensibleObject
objectclass: nsMappingTree
nsslapd-state: backend
nsslapd-backend: userData
cn: dc=test,dc=com


2. Add a database for this suffix :
# ldapmodify -a -p 389 -h localhost -D "cn=Directory manager" -w password
dn: cn=userdata,cn=ldbm database,cn=plugins,cn=config
objectclass: extensibleObject
objectclass: nsBackendInstance
nsslapd-suffix: dc=test,dc=com


3. Disable the suffix
# ldapmodify -p 389 -h localhost -D "cn=Directory manager" -w password
dn: cn="dc=test,dc=com",cn=mapping tree,cn=config
changetype: modify
replace: nsslapd-state
nsslapd-state: disabled


Actual results:

# dsctl instance healthcheck
Beginning lint report, this could take a while ...
Checking Backends ...
Error: Operations error - Warning: Operation attempted on a disabled node :
dc=test,dc=com


Expected results:

DS Lint Error: DSBLE0002
-------------------------------------------------------------------------------
-
Severity: HIGH
Affects:
 -- dc=test,dc=com

Details:
-----------
Unable to query the backend.  LDAP error ({'desc': 'Operations error', 'info':
'Warning: Operation attempted on a disabled node : dc=test,dc=com\n'})

Resolution:
-----------
Check the server's error and access logs for more information.


Additional info:
```

### PR DESCRIPTION
**Pull request by mreynolds389**
Bug Description:  

The healthcheck tool was actually crashing when a suffix was disabled.  We also were not correctly processing DSLdapObjects, where we would run all the lint tests even though we only asked to run one specific lint test.

Fix Description:  

Make healthcheck more robust to handle exceptions.  Fix the processing of DSLdapObjects by passing in the lint function name to DSLint().

Also added the health "check" that triggered the issue to the final report so you know which exact test to rerun.

Fixes: https://github.com/389ds/389-ds-base/issues/4159

### CODE REVIEW

**Review by droideck — APPROVED** [review]
LGTM!

**Comment by Firstyear** [comment]
Ack as well :)

File: dirsrvtests/tests/suites/healthcheck/healthcheck_test.py [hunk 1]
```diff
@@ -455,6 +455,36 @@ def test_healthcheck_database_not_initialized(topology_no_sample):
...
+
+    RET_CODE = 'DSBLE0002'
+
+    be_insts = Backends(topology_st.standalone).list()
+    for be in be_insts:
+        be_suffix = ensure_str(be.get_attr_val_utf8_l('nsslapd-suffix')).lower()
+        if be_suffix == DEFAULT_SUFFIX:
+            be.replace("nsslapd-state", "disabled")
+
+    run_healthcheck_and_flush_log(topology_st, topology_st.standalone, RET_CODE, json=False)
```
Comment (droideck) [inline dirsrvtests/tests/suites/healthcheck/healthcheck_test.py]: Okay, sorry, I looked at the wrong test results thinking that everything is passed...

Now I see that this one failed in `run_healthcheck_and_flush_log` function:
```
>           assert topology.logcap.contains(searched_code)
E           AssertionError: assert False
E            +  where False = <bound method LogCapture.contains of <LogCapture (NOTSET)>>('DSBLE0002')
E            +    where <bound method LogCapture.contains of <LogCapture (NOTSET)>> = <LogCapture (NOTSET)>.contains
E            +      where <LogCapture (NOTSET)> = <lib389.topologies.TopologyMain object at 0x7fb96d8e4d00>.logcap
dirsrvtests/tests/suites/healthcheck/healthcheck_test.py:55: AssertionError
```

Maybe it's my setup somehow but please, check if it works for you.

P.S. Fedora 32 - 389-ds-base built from your PR - PYINSTALL=true

Comment (mreynolds389) [reply dirsrvtests/tests/suites/healthcheck/healthcheck_test.py]: It was a bug in the script - fixed!

**Pull request merged**